### PR TITLE
[refactor] initial rewrite to use native async_hooks

### PR DESCRIPTION
**Pull request by Fishrock123**
Includes lots of debug statements.

We should probably wait to merge this until https://github.com/nodejs/node/pull/8531 is merged.

cc @AndreasMadsen

### CODE REVIEW

File: visualizer/index.html [hunk 1]
```diff
@@ -56,7 +56,7 @@
   section#info #stats {
     float: right;
     width: 155px;
-    height: 119px;
+    height: 135px;
```
Comment (Fishrock123) [inline visualizer/index.html]: Needs a better way to fit the info on the screen. My CSS is very rusty but Ideally we'd place the dprof version and whatnot to the left or right of the other info.

File: dprof.js [hunk 2]
```diff
@@ -100,46 +142,66 @@ Node.prototype.rootIntialize = function () {
...
 const stateMap = new Map();
 let currState = root;
 
-function asyncInit(uid, handle, provider, parentUid) {
+function asyncInit(uid, type, parentUid, handle) {
+  // process._rawDebug('init:' + uid, parentUid);
+
+  // Ignore our nextTick for the root duration
+  // TODO(Fishrock123): detect this better.
+  if (uid === 2) return
```
Comment (Fishrock123) [inline dprof.js]: highlighting this

File: dprof.js [hunk 3]
```diff
@@ -29,25 +52,43 @@ function timestamp() {
   return t[0] * 1e9 + t[1];
 }
 
-function Node(handle, stack) {
-  this.name = handle.constructor.name;
+function Node(handle, id, name, stack) {
+  this.name = name;
+  this._id = id;
```
Comment (Fishrock123) [inline dprof.js]: should maybe be `this._uid = uid`

File: dprof.js [hunk 4]
```diff
@@ -100,46 +142,66 @@ Node.prototype.rootIntialize = function () {
...
 
   // add new state node
-  stateMap.set(uid, state.add(handle));
+  // process._rawDebug('>>> UID:', uid)
+  stateMap.set(uid, state.add(handle, uid, type));
 }
 
 function asyncBefore(uid) {
+  // Ignore our nextTick for the root duration
+  // TODO(Fishrock123): detect this better.
```
Comment (AndreasMadsen) [inline dprof.js]: why is `asyncHook.disable()` not enough?

Comment (Fishrock123) [inline dprof.js]: @AndreasMadsen `disable`/`enable` are no longer tree based, and so you will still get the `before`/`after`/`destroy`.

Comment (AndreasMadsen) [inline dprof.js]: `stateMap` can be used as a set of all seen init events. That should allow us to ignore all before, after, and destroy events where the init event was "disabled" away.

Comment (Fishrock123) [inline dprof.js]: @AndreasMadsen so if `state === undefined` skip?

Comment (AndreasMadsen) [inline dprof.js]: Yes. I was originally thinking `stateMap.has(uid)`, but I guess your version is a little faster.

**Comment by Fishrock123** [comment]
Rebased on `master`.

**Comment by AndreasMadsen** [comment]
Great! It would be awesome if you could fix https://github.com/AndreasMadsen/dprof/issues/10 simultaneously. That would also give us a change to evaluate `asyncHooks.currentId()`.

**Comment by Fishrock123** [comment]
@AndreasMadsen are you looking for the conceptual parent or technical one?

As I understand it the `parentId` passed in `init` will be the conceptual one.

**Comment by AndreasMadsen** [comment]
@Fishrock123 the conceptual one, that is at least the current behaviour. It's almost what is implemented here, but I would like to avoid using `currState`. Maybe something like:

```js
const source = parentUid === 1 ? asyncHooks.currentId() : parentUid;
```

By the way, do you have a good source for the `parentUid === 1` detail? From what I can see, it is not documented in the EP, just in some code comments that I don't fully understand.

**Comment by trevnorris** [comment]
@AndreasMadsen 
> By the way, do you have a good source for the `parentUid === 1` detail?

Think I put it in there as a "note". Not as an actual section. Will improve that. Short of it is there are two reserved id's. `0` and `1`.

- `0`: The "void". The currentId is always set to zero after leaving the JS stack. If `parentId` or `currentId()` are ever `0` then there's a bug.

- `1`: "root". This id is assigned to startup or bootstrap execution. If all id's are mapped out properly all of them should lead back to `1`.

The code you have there is basically the same as changes I'm currently hammering out. `parentId` should always be the id of the handle that originated the constructor. Sometimes it will be the same as `currentId()`. Though after I get done with my most current set of changes you should always be able to rely on `parentId`.

File: dprof.js [hunk 5]
```diff
@@ -104,18 +144,10 @@ Node.prototype.rootIntialize = function () {
...
-asyncHook.addHooks({
-  init: asyncInit,
-  pre: asyncBefore,
-  post: asyncAfter,
-  destroy: asyncDestroy
-});
-
 const root = new Node(
-  0, { 'constructor': { name: 'root' } },
+  0,
```
Comment (AndreasMadsen) [inline dprof.js]: According to what @trevnorris said about the parentId, this should be `1`.

Comment (Fishrock123) [inline dprof.js]: Hmmm, true.

**Pull request merged**